### PR TITLE
config to explicitly specify classpath for hadoop container during hadoop ingestion

### PR DESCRIPTION
**Pull request by himanshug**
see the doc update to index.md for further details.

### CODE REVIEW

File: server/src/main/java/io/druid/initialization/Initialization.java [hunk 1]
```diff
@@ -275,6 +278,41 @@ public static URLClassLoader getClassLoaderForExtension(File extension) throws M
...
+  {
+    try {
+      String[] paths = cp.split(File.pathSeparator);
+
+      List<URL> urls = new ArrayList<>();
+      for (int i = 0; i < paths.length; i++) {
+        File f = new File(paths[i]);
+        if ("*".equals(f.getName())) {
+          File parentDir = f.getParentFile();
+          if (parentDir.exists() && parentDir.isDirectory()) {
```
Comment (guobingkun) [inline server/src/main/java/io/druid/initialization/Initialization.java]: isDirectory() also checks exists ircc

**Comment by guobingkun** [comment]
:+1:  after https://github.com/druid-io/druid/pull/2722#discussion_r57369630

File: server/src/main/java/io/druid/initialization/Initialization.java [hunk 2]
```diff
@@ -275,6 +278,41 @@ public static URLClassLoader getClassLoaderForExtension(File extension) throws M
...
+                    return name != null && (name.endsWith(".jar") || name.endsWith(".JAR"));
+                  }
+                }
+            );
+            for (File jar : jars) {
+              urls.add(jar.toURI().toURL());
+            }
+          }
+        } else {
+          urls.add(new File(paths[i]).toURI().toURL());
```
Comment (pjain1) [inline server/src/main/java/io/druid/initialization/Initialization.java]: Is there a need to check for non-jar files here ?

File: docs/content/configuration/index.md [hunk 3]
```diff
@@ -23,6 +23,7 @@ Many of Druid's external dependencies can be plugged in as modules. Extensions c
 |--------|-----------|-------|
 |`druid.extensions.directory`|The root extension directory where user can put extensions related files. Druid will load extensions stored under this directory.|`extensions` (This is a relative path to Druid's working directory)|
 |`druid.extensions.hadoopDependenciesDir`|The root hadoop dependencies directory where user can put hadoop related dependencies files. Druid will load the dependencies based on the hadoop coordinate specified in the hadoop index task.|`hadoop-dependencies` (This is a relative path to Druid's working directory|
+|`druid.extensions.hadoopContainerDruidClasspath`|Hadoop Indexing launches hadoop jobs and this configuration provides way to explicitly set the user classpath for the hadoop job. By default this is computed automatically by druid based on the druid process classpath and set of extensions. However, sometimes you might want to be explicit to resolve dependency conflicts between druid and hadoop.|druid classpath and extensions|
```
Comment (pjain1) [inline docs/content/configuration/index.md]: I think `Default` should be `null` instead of `druid classpath and extensions`

File: server/src/main/java/io/druid/initialization/Initialization.java (hunk 2 above)
Comment (himanshug) [inline server/src/main/java/io/druid/initialization/Initialization.java]: not really, this method behaves same way as java's classpath

File: server/src/main/java/io/druid/initialization/Initialization.java (hunk 1 above)
Comment (himanshug) [inline server/src/main/java/io/druid/initialization/Initialization.java]: thats correct, changed

File: docs/content/configuration/index.md (hunk 3 above)
Comment (himanshug) [inline docs/content/configuration/index.md]: changed

File: docs/content/configuration/index.md [hunk 4]
```diff
@@ -23,6 +23,7 @@ Many of Druid's external dependencies can be plugged in as modules. Extensions c
 |--------|-----------|-------|
 |`druid.extensions.directory`|The root extension directory where user can put extensions related files. Druid will load extensions stored under this directory.|`extensions` (This is a relative path to Druid's working directory)|
 |`druid.extensions.hadoopDependenciesDir`|The root hadoop dependencies directory where user can put hadoop related dependencies files. Druid will load the dependencies based on the hadoop coordinate specified in the hadoop index task.|`hadoop-dependencies` (This is a relative path to Druid's working directory|
+|`druid.extensions.hadoopContainerDruidClasspath`|Hadoop Indexing launches hadoop jobs and this configuration provides way to explicitly set the user classpath for the hadoop job. By default this is computed automatically by druid based on the druid process classpath and set of extensions. However, sometimes you might want to be explicit to resolve dependency conflicts between druid and hadoop.|null|
```
Comment (fjy) [inline docs/content/configuration/index.md]: can you add some details about how to potentially use this in other hadoop doc?

Comment (himanshug) [inline docs/content/configuration/index.md]: @fjy added

File: server/src/main/java/io/druid/initialization/Initialization.java [hunk 5]
```diff
@@ -275,6 +278,41 @@ public static URLClassLoader getClassLoaderForExtension(File extension) throws M
     return loader;
   }
 
+  public static List<URL> getURLsForClasspath(String cp)
+  {
+    try {
+      String[] paths = cp.split(File.pathSeparator);
```
Comment (drcrallen) [inline server/src/main/java/io/druid/initialization/Initialization.java]: Is it possible to use Path here instead of String arrays?

File: server/src/main/java/io/druid/initialization/Initialization.java [hunk 6]
```diff
@@ -275,6 +278,41 @@ public static URLClassLoader getClassLoaderForExtension(File extension) throws M
...
+
+      List<URL> urls = new ArrayList<>();
+      for (int i = 0; i < paths.length; i++) {
+        File f = new File(paths[i]);
+        if ("*".equals(f.getName())) {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/initialization/Initialization.java]: My concern here is that the classpath parsing is not done using java internal parsing stuff, so someone could have a classpath as an argument for their java command line, and specify some other classpath for the argument here, and end up with different sets of stuff.

Is there a way this classpath parsing can be done using the same parsing as the jvm?

Comment (himanshug) [inline server/src/main/java/io/druid/initialization/Initialization.java]: if you mean using something from jvm to convert provided classpath string to List of URL, then I don't think there is any standard method in the jdk to do that. Please let me know if there is any and I'll remove this.

btw it is intentional that user will be able to configure different classpath for druid and for the hadoop containers.

Comment (himanshug) [inline server/src/main/java/io/druid/initialization/Initialization.java]: also, this PR has been tested on one of our clusters.

File: docs/content/ingestion/batch-ingestion.md [hunk 7]
```diff
@@ -82,6 +82,8 @@ instance of a Druid [overlord](../design/indexing-service.html). A sample task i
 |hadoopDependencyCoordinates|A JSON array of Hadoop dependency coordinates that Druid will use, this property will override the default Hadoop coordinates. Once specified, Druid will look for those Hadoop dependencies from the location specified by `druid.extensions.hadoopDependenciesDir`|no|
 |classpathPrefix|Classpath that will be pre-appended for the peon process.|no|
 
+also note that, druid automatically computes the classpath for hadoop job containers that run in hadoop cluster. But, in case of conflicts between hadoop and druid's dependencies, you can manually specify the classpath by setting `druid.extensions.hadoopContainerDruidClasspath` property. See the extensions config in [base druid configuration](../configuration/index.html).
```
Comment (fjy) [inline docs/content/ingestion/batch-ingestion.md]: This should go in the other-hadoop.md file. Also, I am confused when I should override this property and when I should keep the default

Comment (himanshug) [inline docs/content/ingestion/batch-ingestion.md]: other-hadoop.md is about setting up a different hadoop and unrelated to this. this is to be used when your MR containers failing due to a conflict in druid dependencies and hadoop's own dependencies (or if you have hdfs-storage and hadoop jars on main druid process classpath like we do leading to duplicate versions of jersey, guava etc showing up... our hadoop versions do not live in mvn or any other conflicts). 
basically when druid computed classpath for hadoop job does not work for you for whatever reason, you can override same by this setting.

**Comment by fjy** [comment]
:+1:

File: server/src/main/java/io/druid/initialization/Initialization.java (hunk 6 above)
Comment (drcrallen) [inline server/src/main/java/io/druid/initialization/Initialization.java]: Yeah, I couldn't find a good standard way when i peeked around. For "normal" looking classpaths this shouldn't be an issue.
